ros2 pkg create

Karsten1987 opened the pull request:
open PR for visibility and early feedback. It does work, compile etc, but is completely untested and most likely has massive style issues when it comes to python :)

`ros2 pkg create` allows to create dummy ros2 packages.

example:
```
ros2 pkg create my_package --dependencies rclcpp rclpy --create-cpp-exe
```
```
 ➭ tree my_package/
my_package/
├── CMakelists.txt
├── package.xml
└── src
    └── main.cpp

1 directory, 3 files
```

``` package.xml
 ➭ cat my_package/package.xml
<?xml version="1.0"?>
<?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
<package format="2">
  <name>my_package</name>
  <version>0.0.0</version>
  <description>Description of my_package</description>
  <maintainer email="karsten@karsten-macbook-pro.local">karsten</maintainer>
  <license>Apache License 2.0</license>

  <buildtool_depend>cmake</buildtool_depend>

  <build_depend>rclcpp</build_depend>
  <build_depend>rclpy</build_depend>

  <exec_depend>rclcpp</exec_depend>
  <exec_depend>rclpy</exec_depend>

  <export>
    <build_type>cmake</build_type>
  </export>
</package>
```
``` CMakeLists.txt
 ➭ cat my_package/CMakelists.txt
cmake_minimum_required(VERSION 3.5)
project(my_package)

find_package(rclcpp REQUIRED)
find_package(rclpy REQUIRED)


include_directories(
  rclcpp_INCLUDE_DIRS
  rclpy_INCLUDE_DIRS

)

add_executable(${PROJECT_NAME} src/main.cpp)
target_link_libraries(${PROJECT_NAME} ${rclcpp_LIBRARIES})
target_link_libraries(${PROJECT_NAME} ${rclpy_LIBRARIES})

install(TARGETS ${PROJECT_NAME}
  DESTINATION lib/${PROJECT_NAME})


ament_package()
```

mikaelarguedas commented:
That's a great initiative! :+1: 

For reference there is interest or related work being done in [catkin_pkg](https://github.com/ros-infrastructure/catkin_pkg/pull/133) and [catkin_tools](https://github.com/catkin/catkin_tools/issues/292).

As of the content of the template and the options, I think that we can do many things and will have to decide what to target. According to the feature set we settle on, it will guide the final location of this executable and the extend to which it can be shared with the ROS1 ecosystem

dirk-thomas commented on ros2pkg/ros2pkg/command/pkg.py:
```diff
@@ -29,6 +29,5 @@ def main(self, *, parser, args):
         # the attribute should always exist
         # otherwise argparse should have exited
         extension = getattr(args, '_verb')
-
```
Please avoid unrelated whitespace changes.

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
```
I would suggest to call the file just `create` since `api` is already part of the namespace.

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+                  (template_file, output_file), file=sys.stderr)
+            raise
+    content = output.getvalue()
+    interpreter.shutdown()
```
This line should be moved into a `finally` block so that it always happens - even in case of an exception.

Karsten1987 replied:
I introduced a `managed interpreter` which calls `shutdown()` in the `__exit__` function. Is that a correct python way of doing?

dirk-thomas replied:
Since it is only used in one location that is quite some block of code (without three lines `try`, `finally`, and `shutdown` would be enough). But it is a "correct" way in Python. Anyone reading the code will have to look up what this custom class does rather than directly understand the code where it is being used. So I would slightly lean to avoid a custom class it in this case.

Beside that I would argue though that you shouldn't embed the constructor arguments into the custom class since it prevents reuse. Also the constructor signature of your custom class is different from the base class which is confusing to the user imo.

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+    else:
+        # create folder if necessary
+        try:
+            os.makedirs(os.path.dirname(output_file))
```
`os.makedirs` has an optional argument `exist_ok` which can be used to avoid the try/except.

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+
+def create_folder(folder_name, base_directory):
+    if not os.path.isabs(base_directory):
+        print("cannot create %s, folder path is not absolute" % base_directory)
```
Please put single quotes around the `%s`.

Same below.

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+        return False
+
+    folder_path = os.path.join(base_directory, folder_name)
+    if (os.path.exists(folder_path)):
```
No need to wrap the condition in parenthesis.

Same below.

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+        print("cannot create %s, folder exists" % folder_path)
+        return False
+
+    print("creating folder", folder_path)
```
Please use single quotes instead of double quotes (where possible without the need for escaping).

Same below.

dirk-thomas commented on ros2pkg/setup.py:
```diff
@@ -35,6 +35,7 @@
             'executables = ros2pkg.verb.executables:ExecutablesVerb',
             'list = ros2pkg.verb.list:ListVerb',
             'prefix = ros2pkg.verb.prefix:PrefixVerb',
+            'create = ros2pkg.verb.create:CreateVerb',
```
Alpha order.

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+    template_path = os.path.abspath(
+        os.path.join(os.path.realpath(__file__), '../../resource/'+template_file_name))
+    if not os.path.exists(template_path):
+        print('template not found: ', template_path)
```
The printed line will have two spaces between the `:` and the path.

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+        os.path.join(os.path.realpath(__file__), '../../resource/'+template_file_name))
+    if not os.path.exists(template_path):
+        print('template not found: ', template_path)
+        return False
```
The caller doesn't check for the return value. Also this shouldn't be happening therefore I would suggest an exception.

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+
+def create_template_file(template_file_name, output_directory, output_file_name, template_config):
+    template_path = os.path.abspath(
+        os.path.join(os.path.realpath(__file__), '../../resource/'+template_file_name))
```
The base path for the templates should be extracted into a module-level constant. (That e.g. allows tests to mock them easily.)

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+        h.write(content)
+
+def create_folder(folder_name, base_directory):
+    if not os.path.isabs(base_directory):
```
Wouldn't the code also work with relative paths? Why restrict it to absolute?

dirk-thomas commented on ros2pkg/ros2pkg/api/create_api.py:
```diff
@@ -0,0 +1,87 @@
+    with open(output_file, 'w') as h:
+        h.write(content)
+
+def create_folder(folder_name, base_directory):
```
What does this function offer beyond `os.makedirs(os.path.join(base_directory, folder_name))`?

dirk-thomas commented on ros2pkg/ros2pkg/resource/CMakeLists.txt.em:
```diff
@@ -0,0 +1,19 @@
+)
+
+@[if create_cpp_exe]add_executable(${PROJECT_NAME} src/@(cpp_exe_name))
+@[for deb in dependencies]target_link_libraries(${PROJECT_NAME} ${@(deb)_LIBRARIES})
```
Instead of `include_directories` and `target_link_libraries` this should use `ament_target_dependencies`.

dirk-thomas commented on ros2pkg/ros2pkg/resource/CMakeLists.txt.em:
```diff
@@ -0,0 +1,19 @@
+project(@(project_name))
+
+@[for deb in dependencies]find_package(@deb REQUIRED)
+@[end for]
```
For ease of readability this should be formatted like this:

```
@[for deb in dependencies]@
find_package(@deb REQUIRED)
@[end for]@
```

Note the trailing `@` to avoid a newline.

Same below.

dirk-thomas commented on ros2pkg/ros2pkg/resource/main.cpp.em:
```diff
@@ -0,0 +1,7 @@
+
+int main(int argc, char ** argv)
+{
+  std::cout << "hello @(package_name)" << std::endl;
```
Since we don't recommend using `cout` it shouldn't be used in the generated code.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,112 @@
+
+from ros2pkg.verb import VerbExtension
+from ros2pkg.api.create_api import create_folder
+from ros2pkg.api.create_api import create_template_file
```
Alpha import order.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,112 @@
+from ros2pkg.api.create_api import create_template_file
+
+class CreateVerb(VerbExtension):
+    """Output the prefix path of a package."""
```
Copy-n-paste.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,112 @@
+            help='Directory where to create the package')
+        arg = parser.add_argument(
+            '--build-tool',
+            default='cmake',
```
Isn't `ament_cmake` (not `ament-cmake`) the more common build type?

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,112 @@
+            '--build-tool',
+            default='cmake',
+            choices=['cmake', 'ament-cmake'],
+            help='Which build tool to use')
```
"build tool" is the wrong term here. "The build type of the package" would be more accurate.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,112 @@
+            help='list of dependencies')
+        arg = parser.add_argument(
+            '--maintainer-email',
+            default=getpass.getuser()+'@'+os.uname()[1]+'.local',
```
PEP8: spaces around operators.

Same above.

The dafault value seems to be not helpful since it is unlikely that it is a valid email address.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,112 @@
+            '--create-cpp-exe',
+            nargs='?',
+            const=True,
+            required=False,
```
This argument should use `action='store_true'` instead.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,112 @@
+            '--cpp-exe-name',
+            nargs='+',
+            default='main.cpp',
+            help='name of the empty cpp executable')
```
The name and help seems a bit misleading. The arguments are the source filenames - not the executable.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,112 @@
+        if (args.create_cpp_exe):
+            src_folder = create_folder('src', package_directory)
+            if not src_folder:
+                return
```
Should this really fail silently?

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,112 @@
+            'maintainer_email' : maintainer_email,
+            'maintainer_name' : maintainer_name,
+            'dependencies' : dependencies,
+            }
```
PEP8: The closing parenthesis shouldn't be indented.

Karsten1987 commented:
thanks for the feedback. I am sorry, I didn't know about the existing work in ROS1.
I wanted to make sure that an approach like this would make sense and if this is the right place for it. I'll iterate over it then, correct it in style and write tests for it.

dirk-thomas commented:
> I wanted to make sure that an approach like this would make sense and if this is the right place for it.

I think the approach looks good and the place to. `ros2 pkg create` makes sense.

Karsten1987 commented:
I hopefully addressed the comments mentioned above. I put this up for review, noting that I won't have much more time before beta3 to work on it. 

Currently available:
- ament_cmake package configurable with dependencies
- plain cmake package

Currently missing:
- cmake config file configuration
- python packages
- verbose tests

Now, I am fine with either, holding this back and only merge it when at least completed with tests or mark it as experimental (or something) to get it merged as is.

With that in mind, the CI:
linux: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3197)](http://ci.ros2.org/job/ci_linux/3197/)
win: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=3259)](http://ci.ros2.org/job/ci_windows/3259/)
osx: to come when machines are back online

dirk-thomas commented:
Please don't run any further CI builds for this. Since this patch doesn't add any tests or anything which uses the new code (not even a `SyntaxError` in the Python code would be catched) they are useless.

dirk-thomas commented:
Nvm, the tests will at least do the style checking. Makes total sense.

dirk-thomas commented:
But running an OS X build is not necessary since the style check will be the same as on any other platform.

dirk-thomas commented on ros2pkg/ros2pkg/api/create.py:
```diff
@@ -0,0 +1,94 @@
+import em
+
+
+RELATIVE_RESOURCE_PATH = '../../resource/'
```
The relative path is misleading since it goes two levels up which only works if the first level is actually the filename itself. Also paths should have a trailing slash.

I would change this to `RESOURCE_PATH` and actually store the following location: `os.path.join(os.path.dirname(__file__), '..', 'resource')`.

No need for `realpath` here.

dirk-thomas commented on ros2pkg/ros2pkg/api/create.py:
```diff
@@ -0,0 +1,94 @@
+RELATIVE_RESOURCE_PATH = '../../resource/'
+
+
+def expand_template(template_file, data, output_file, minimum_timestamp=None):
```
I don't see why this use case would need the `minimum_timestamp` logic. Since it doesn't seem to be used it should be removed.

dirk-thomas commented on ros2pkg/ros2pkg/api/create.py:
```diff
@@ -0,0 +1,94 @@
+def create_folder(folder_name, base_directory):
+    # if not os.path.isabs(base_directory):
+    #     print("cannot create '%s', folder path is not absolute" % base_directory)
+    #     return False
```
What about this commented code?

dirk-thomas commented on ros2pkg/ros2pkg/api/create.py:
```diff
@@ -0,0 +1,94 @@
+
+    folder_path = os.path.join(base_directory, folder_name)
+    if os.path.exists(folder_path):
+        print("cannot create '%s', folder exists" % folder_path)
```
To `stderr`.

dirk-thomas commented on ros2pkg/ros2pkg/api/create.py:
```diff
@@ -0,0 +1,94 @@
+    template_path = os.path.abspath(
+        os.path.join(os.path.realpath(__file__), RELATIVE_RESOURCE_PATH + template_file_name))
+    if not os.path.exists(template_path):
+        raise IOError('template not found:', template_path)
```
Could this raise `FileNotFoundError` instead?

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,82 @@
+@[else]@
+# uncomment the following section in order to fill in
+# further dependencies manually.
+# find_package(<dependency> [<REQUIRED>|<QUIET>])
```
`REQUIRED` and `QUIET` can both be used at the same time.

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,82 @@
+@[if dependencies or create_cpp_library]@
+include_directories(
+@[if create_cpp_library]@
+  include
```
it would be good if the script would also create the directory `include/<pkgname>` in that case (maybe even a header?). Otherwise using the package might result in warning since the include path doesn't exist.

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,82 @@
+# find dependencies
+find_package(ament_cmake REQUIRED)
+@[if dependencies]@
+@[for deb in dependencies]@
```
Please indent nested `em` commands: `@[  for deb in dependencies]@`.

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,82 @@
+find_package(@deb REQUIRED)
+@[end for]@
+
+@[if dependencies or create_cpp_library]@
```
This is within the block `@[if dependencies]@` so the condition doesn't seem to make sense?

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,82 @@
+if(BUILD_TESTING)
+  find_package(ament_lint_auto REQUIRED)
+  ament_lint_auto_find_test_dependencies()
+endif()
```
I would recommend all testing related code to the very end of the CMake file (even after `ament_package()`) in order to ensure that the test related code doesn't affect the package installation.

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/package.xml.em:
```diff
@@ -0,0 +1,29 @@
+@[end for]@
+
+@[for deb in dependencies]@
+  <exec_depend>@deb</exec_depend>
```
You could use the `depend` tag instead of `build_depend` and `exec_depend`.

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/package.xml.em:
```diff
@@ -0,0 +1,29 @@
+@[end for]@
+
+@[end if]@
+  <test_depend>ament_cmake_gtest</test_depend>
```
The CMake template doesn't seem to use any `gtest`?

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,82 @@
+@[end if]@
+@[end if]@
+@[if create_cpp_library]@
+add_library(${PROJECT_NAME} src/@(cpp_library_name))
```
I would suggest to wrap before the source file to encourage users to list them on separate lines when the list grows.

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,82 @@
+
+@[end if]@
+@[if create_cpp_exe]@
+add_executable(${PROJECT_NAME}_main src/@(cpp_exe_name))
```
The naming of the executable with the `_main` suffix might be undesired. I don't think this is a good starting point for a new user.

dirk-thomas commented on ros2pkg/ros2pkg/resource/cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,99 @@
+  DESTINATION lib/${PROJECT_NAME})
+
+@[end if]@
+if(BUILD_TESTING)
```
The CMake option `BUILD_TESTING` is not defined in plain CMake when not including `CTest` explicitly.

dirk-thomas commented on ros2pkg/ros2pkg/resource/cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,99 @@
+
+# Export the package for use from the build-tree
+# (this registers the build-tree with a global CMake-registry)
+export(PACKAGE @(project_name))
```
I don't think this is good practice. We don't want other packages to rely on this.

dirk-thomas commented on ros2pkg/ros2pkg/resource/cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,99 @@
+set(export_targets ${export_targets};${PROJECT_NAME}_main)
+@[end if]@
+export(TARGETS ${export_targets}
+  FILE "${PROJECT_BINARY_DIR}/@(project_name)Targets.cmake")
```
The generated targets file needs to be installed.

dirk-thomas commented on ros2pkg/ros2pkg/resource/cmake/Config.cmake.in.em:
```diff
@@ -0,0 +1,16 @@
+#  @(project_name)_EXECUTABLE   - the bar executable
+ 
+# Compute paths
+get_filename_component(@(project_name)_CMAKE_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)
```
No need to create your own variable for this. `@(PROJECT_NAME)_DIR` is defined automatically pointing to the directory containing the CMake config file of the package.

dirk-thomas commented on ros2pkg/ros2pkg/resource/cmake/Config.cmake.in.em:
```diff
@@ -0,0 +1,16 @@
+
+# These are IMPORTED targets created by @(project_name)Targets.cmake
+set(@(project_name)_LIBRARIES foo)
+set(@(project_name)_EXECUTABLE bar)
```
Where is `foo` and `bar` coming from?

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,179 @@
+            help='The package name')
+        parser.add_argument(
+            '--destination-directory',
+            default=os.path.realpath(os.getcwd()),
```
No need for `realpath` here. The default could just be `os.curdir`.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,179 @@
+            help='list of dependencies')
+        parser.add_argument(
+            '--maintainer-email',
+            default=getpass.getuser() + '@' + platform.uname()[1] + '.local',
```
Using `.local` as a default is a bad idea imo. Based on experience too many users will keep it as is.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,179 @@
+        parser.add_argument(
+            '--cpp-exe-name',
+            nargs='+',
+            help='name of the empty cpp executable')
```
Why need two options `--cpp-exe-name` and `--create-cpp-exe`. One optional list of executable names should be sufficient. If it is empty the user doesn't want any executable.

Same for libraries.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,179 @@
+        build_tool = args.build_tool
+        dependencies = args.dependencies
+        maintainer_email = args.maintainer_email
+        maintainer_name = args.maintainer_name
```
All these local variables seem unnecessary. Just using the `args.*` variables should be ok or is there a reason not to?

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,179 @@
+
+        package_directory = create_folder(package_name, destination_directory)
+        if not package_directory:
+            return
```
There is no error code for this case. The caller can't determine that the script failed.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,179 @@
+            'dependencies': dependencies,
+        }
+
+        if (build_tool == 'cmake'):
```
Unnecessary parenthesis.

mikaelarguedas commented on ros2pkg/ros2pkg/resource/cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,99 @@
+
+@[end if]@
+@[if create_cpp_library]@
+add_library(${PROJECT_NAME} src/@(cpp_library_name))
```
in the case of pure cmake packages they need to be built as shared libraries or to add the "BUILD_SHARED" option

Karsten1987 commented:
Hopefully addressed all comments so far. Both, ament_cmake and plain cmake, work with the current version. Tested it by creating a second package manually, using the first package and linking against it.

When creating a library, it also creates the include folder with two headers (including the visibility macros for win). 
If no name/email is given, I'll try to fetch them from a global git config. If this also fails, I'll set the email to `<current_user>@todo.todo` as specified in ROS1.

dirk-thomas commented on ros2pkg/ros2pkg/resource/cpp/header.hpp.em:
```diff
@@ -0,0 +1,19 @@
+#ifndef @(package_name.upper())__@(library_name.upper())__HPP_
```
I don't think this will pass linting. There should be only a single underscore before `HPP`.

dirk-thomas commented on ros2pkg/ros2pkg/resource/cpp/header.hpp.em:
```diff
@@ -0,0 +1,19 @@
+public:
+  @(class_name)();
+
+  virtual ~@(class_name)() {}
```
Should the destructor be defined in the `.cpp` file instead?

dirk-thomas commented on ros2pkg/ros2pkg/api/create.py:
```diff
@@ -0,0 +1,86 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
```
Nitpick: no need for two lines. Same for line 21/22.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,225 @@
+            default=os.curdir,
+            help='Directory where to create the package')
+        parser.add_argument(
+            '--build-tool',
```
The correct term is "build type".

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,225 @@
+        parser.add_argument(
+            '--destination-directory',
+            default=os.curdir,
+            help='Directory where to create the package')
```
Maybe "Directory where to create the package directory" is more accurate (assuming this is what is happening)?

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,225 @@
+import getpass
+import os
+import subprocess
+from subprocess import PIPE
```
Since `subprocess` is already imported I would suggest to remove this line and use `subprocess.PIPE` instead.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,225 @@
+
+        maintainer_name = getpass.getuser()
+        if not args.maintainer_name:
+            # try getting the name from the global git config
```
Is the git username a good fallback? I would expect that it almost never has a "normal" firstname + lastname.

Karsten1987 replied:
I think it's valid fallback. At least the information inside the package.xml corresponds to github commits (if any).

Do you have any ideas?

dirk-thomas replied:
The current username of the system could be one. But maybe also potentially not useful...

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,225 @@
+        print('maintainer_name:', maintainer_name)
+        print('dependencies:', args.dependencies)
+        print('cpp_node_name:', args.cpp_node_name)
+        print('cpp_library_name:', args.cpp_library_name)
```
Some args might be `None`. Maybe it would be nicer to not print a line in that case?

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,225 @@
+
+        package_directory = create_folder(args.package_name, args.destination_directory)
+        if not package_directory:
+            print('unable to create folder', args.destination_directory)
```
To `stderr`.

Same below.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,225 @@
+
+            cmake_config = {
+                'project_name': args.package_name,
+                'create_cpp_library': True if args.cpp_library_name else False,
```
The expression can be simplified to: `bool(args.cpp_library_name)`.

Same below.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,225 @@
+        package_directory = create_folder(args.package_name, args.destination_directory)
+        if not package_directory:
+            print('unable to create folder', args.destination_directory)
+            return False
```
I think the return value semantic is different. Please check the return code of the command invocation in these cases. I think you can skip the print line and simply `return 'unable to create folder: ' + args.destination_directory` here.

Same below.

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,225 @@
+                'visibility_control.h',
+                visibility_config)
+
+        return True
```
Related to the previous comment. I think this line needs to be removed (so that the function return `None` in case of success).

Karsten1987 commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,104 @@
+  ament_uncrustify()
+  # the lines above can be replaced with the lines below
+  # but it requires this repo to be a git repo in order
+  # to run all tests correctly
```
when not a git repo, cpplint complains about the header guards not being format conform. It somehow interprets the full path of the file. When calling `git init` inside the package. THe headers are correct.
```
/Users/karsten/workspace/osrf/ros2_full/src/my_package/include/my_package/my_library.hpp:1:  #ifndef header guard has wrong style, please use: __USERS__KARSTEN__WORKSPACE__OSRF__ROS2_FULL__SRC__MY_PACKAGE__INCLUDE__MY_PACKAGE__MY_LIBRARY_HPP_  [build/header_guard] [5]
```

furthermore, copyright checks would also fail.

dirk-thomas replied:
I am not sure what it would take to makes these linters work "outside" of git. But I would suggest to still use ament auto and only disable the linters which have that problem with a comment mentioning why (see https://github.com/ros2/kdl_parser/blob/4151131ddd3339ab93bbf40bbe9cbc4fe1c6bc41/kdl_parser/CMakeLists.txt#L56 for an example). That way all other linters are still be used - including the ones potentially being added in the future.

Karsten1987 commented on ros2pkg/ros2pkg/resource/ament_cmake/package.xml.em:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="2">
```
Do we have a url to the xsd ?

dirk-thomas replied:
The URL will be the same as the one for format, just substituting 2 with 3. The file hasn't been uploaded yet since the REP is not "final" yet.

dirk-thomas replied:
I kind of changes my mind on this. As long as we don't use any format 3 features I would keep it at two (which has a schema available).

Karsten1987 commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,104 @@
+if(NOT CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 14)
+endif()
+
```
done in line 14-16

dirk-thomas commented on ros2pkg/ros2pkg/verb/create.py:
```diff
@@ -0,0 +1,230 @@
+            help='name of the maintainer of this package'),
+        parser.add_argument(
+            '--cpp-node-name',
+            nargs='?',
```
Why is the node name optional? If I pass only `--cpp-node-name` no executable gets generated.

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,98 @@
+)
+@[end if]@
+@[if cpp_node_name]@
+add_executable(${PROJECT_NAME}_node src/@(cpp_node_name))
```
When invoking the `create` verb with `--cpp-node-name baz` I would expect to get an executable with the name I passed in - not `baz_node` in this case.

dirk-thomas replied:
If creating both a library and an executable it would be a good default if the executable target would link to the library?

dirk-thomas commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,98 @@
+  include
+)
+
+add_library(${PROJECT_NAME} src/@(cpp_library_name))
```
Similar to the comment about the executable name. I would expect to get a target with the name passed in with `--cpp-library-name`.

wjwwood commented on ros2pkg/ros2pkg/resource/ament_cmake/CMakeLists.txt.em:
```diff
@@ -0,0 +1,98 @@
+  find_package(ament_lint_auto REQUIRED)
+  # the following line skips the linter which checks for copyrights
+  # remove the line when a copyright and license is present in all source files
+  set(ament_cmake_copyright_FOUND TRUE)
```
Is this really the best way to disable these? @dirk-thomas?

You could also just comment out the dependency in the `package.xml` right?

wjwwood commented on ros2pkg/ros2pkg/resource/ament_cmake/package.xml.em:
```diff
@@ -0,0 +1,24 @@
+
+@[end if]@
+  <test_depend>ament_lint_auto</test_depend>
+  <test_depend>ament_common</test_depend>
```
Is `ament_common` a thing? Or is it `ament_lint_common`?

Maybe it would be better to list them individually, so it's easier to turn them on/off by just removing them or commenting them in the `package.xml`?

dirk-thomas replied:
> Is `ament_common` a thing? Or is it `ament_lint_common`?

It should indeed be [ament_lint_common](https://github.com/ament/ament_lint/tree/master/ament_lint_common).

> Maybe it would be better to list them individually, so it's easier to turn them on/off by just removing them or commenting them in the `package.xml`?

The advantage of depending on the metapackage is that the package will automatically benefit from linters being added in the future.

wjwwood replied:
Dunno, for our packages I think automatically getting new tests is ok, but if I were a maintainer of some higher level package, I'd probably prefer to opt into new linters rather than have them added under my feet. Either way we need a better way to opt-in/opt-out.

dirk-thomas commented:
The functions currently in `ros2pkg/api/create.py` are not really an API to create packages. I would think these functions should all by "private" / starting with an underscore.

The actual logic is actually in `ros2pkg/verb/create.py` and should imo be moved into the first file.

The general idea is that the `api` folder provides API to get the actual task done. The `verb` folder only contains the necessary CLI interface and should call the API for any actual work (beside command line argument checking and mapping to function parameters). So any external code can call the API to achieve the same thing.

Karsten1987 commented:
i changed it to the ament package API after @wjwwood first approval. So I putting that up for another review.

mikaelarguedas commented on ros2pkg/ros2pkg/resource/ament_cmake/package.xml.em:
```diff
@@ -0,0 +1,24 @@
+  <version>0.0.0</version>
+  <description>Description of @package_name</description>
+  <maintainer email="@maintainer_email">@maintainer_name</maintainer>
+  <license>Apache License 2.0</license>
```
as the command line tool takes a license: the user provided license should be expanded here rather than hard-coding Apache 2 (same in the other template below)

mikaelarguedas commented on ros2pkg/ros2pkg/resource/ament_cmake/package.xml.em:
```diff
@@ -0,0 +1,24 @@
+  <maintainer email="@maintainer_email">@maintainer_name</maintainer>
+  <license>Apache License 2.0</license>
+
+  <buildtool_depend>ament_cmake</buildtool_depend>
```
This should depend on `ament_cmake_ros` if a `--cpp-library-name` is passed

mikaelarguedas commented on ros2pkg/ros2pkg/resource/ament_cmake/package.xml.em:
```diff
@@ -0,0 +1,24 @@
+<package format="2">
+  <name>@package_name</name>
+  <version>0.0.0</version>
+  <description>Description of @package_name</description>
```
Should this include the description provided by the user and fall back to this if None provided? (maybe this is planned for a follow up PR)

mikaelarguedas commented:
I don't want to delay the merging of this and didnt actually review it. I just tried the tool and noted some things that surprised me in the resulting package. The only things that I think should be addressed (and easy to address) are the license in the package.xml and the missing `ament_cmake_ros` dep.

Karsten1987 commented:
@mikaelarguedas Thanks a lot for your comments. Would you mind giving it another shot to see if your concerns were addressed?

running CI for linters: [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4041)](https://ci.ros2.org/job/ci_linux/4041/)

dhood commented:
FYI, while testing "fat archive" binaries on a non-dev machine I've discovered that there are runtime dependencies brought in by this PR that were missing from the from-binary install instructions: `python3-empy` and `python3-pyparsing` (indirectly through `ament_package`).

Wiki updates:
- [linux](https://github.com/ros2/ros2/wiki/Linux-Install-Binary/_compare/e5702a53abbc62aecf1c4a02af87e17b96d8b674...088133be89d54f7c62d15e3e8c40bde6c491517b)
- [osx](https://github.com/ros2/ros2/wiki/OSX-Install-Binary/_compare/b67511b9323d016e0993e5eb2a32f52b50311208...72ef8ac80f8d7e0fbe3e49dc03c745291c505362)
- [windows](https://github.com/ros2/ros2/wiki/Windows-Install-Binary/_compare/9cdd189c494b0cb543cc9153da10fb5a818241f3...f7ccfaaace5a8d3de81dae89cb4bcf590cce106c)

dirk-thomas commented:
@Karsten1987 can you follow up on this and add the necessary run dependencies to the packages. It looks like at least `ament_package` isn't mentioned in the manifest.

wjwwood commented:
Retroactive summary for this feature:

add ros2 pkg create
==============

### Purpose / Use Cases (Why implement this feature?)

This was implemented to provide a way to generate a new package from a template, making it easier to create new packages by giving some starting files and structure.

### Design (How does it work?)

It provides a new "verb" to the `ros2 pkg` command line tool called `create` which allows you to generate a new package locally with some common variables passed to the tool and templated into the new package files.

#### Inputs / Outputs / API (How to use it?)  

The usage of the tool is documented in the help text with `ros2 pkg create -h`.

There is a "public" API which implements the actions the tool can take (used internally by the tool), see:

https://github.com/ros2/ros2cli/blob/211fc1648c03e1f4dbf08910ec3f96be66ecca2e/ros2pkg/ros2pkg/api/create.py#L1
